### DISCUSSION TITLE
Google's Perch v2 Bird Vocalization Classifier Released

### DISCUSSION THREAD

**Discussion by tphakala** [tphakala/birdnet-go#1110] [General]
Google Research has just released Perch v2, a significant advancement in bird vocalization classification that could greatly benefit the BirdNET-Go

  What is Perch v2?

  Perch v2 is Google's latest bird vocalization classifier with impressive capabilities:

  - 🌍 10,000+ bird species (vs BirdNET's 6,000)
  - 📊 Trained on complete Xeno-Canto corpus - massive global dataset
  - 🧠 1280-dimensional embeddings for advanced ML applications
  - ⚖️ Apache 2.0 license - fully open for commercial use
  - 🔬 State-of-the-art accuracy on multiple benchmark datasets

  Key Resources

  - https://www.kaggle.com/models/google/bird-vocalization-classifier/tensorFlow2/perch_v2
  - https://tfhub.dev/google/bird-vocalization-classifier/1
  - https://github.com/google-research/perch

  Integration Opportunity

  This represents a major opportunity to enhance BirdNET-Go with cutting-edge bird detection capabilities. I've created a detailed feature request exploring the integration possibilities:

  🔗 https://github.com/tphakala/birdnet-go/issues/1109

  Technical Challenge

  The main challenge is integrating TensorFlow models (Perch v2 uses SavedModel format) alongside our current TFLite models. This would require:

  - TensorFlow C API integration via CGO
  - 32kHz audio preprocessing (vs current 48kHz)
  - Model format abstraction
  - Embedding extraction capabilities

  Call for Contributors 🙋‍♂️

  This is a substantial technical undertaking that would benefit greatly from community collaboration. Are you interested in helping bring Perch v2 to BirdNET-Go?

  Particularly valuable skills:
  - ✅ TensorFlow C API / CGO experience
  - ✅ Model conversion expertise (SavedModel ↔ TFLite/ONNX)
  - ✅ Audio signal processing
  - ✅ Go systems programming

  Even if you don't have direct ML experience, there are many ways to contribute - testing, documentation, benchmarking, etc.

  Discussion Points

  What are your thoughts on:

  1. Priority level - How important is expanded species coverage to you?
  2. Implementation approach - TensorFlow C API vs model conversion vs hybrid service?
  3. Binary size trade-offs - Acceptable increase for 4,000+ additional species?
  4. Use cases - What would you do with 1280-dim bird embeddings?

  The Vision

  Imagine BirdNET-Go with:
  - 🎯 Dual model support - Choose BirdNET or Perch v2 per use case, or use both providing higher accuracy
  - 🔍 Embedding-based search - "Find birds similar to this recording"
  - 📈 Model ensemble - Combine predictions for higher accuracy
  - 🌐 Global coverage - Best species support worldwide

  This could position birdnet-go as the definitive tool for bird vocalization analysis!

 ---
Ready to make BirdNET-Go even more awesome? Check out https://github.com/tphakala/birdnet-go/issues/1109 and let's discuss how to move forward! 🚀

**Comment by cegan09** [comment]
I have none of the skills that you're asking for, but if I can help, I would like to help. I don't have formal software training, I'm a hardware guy, but I'm self taught just enough that me + an AI chatbot is dangerous. I managed to get the new perch V2 setup locally, and I'm able to throw processed audio chunks from my various rtsp streams at it and get results from its built in classifications. I've not done any deep comparisons to how it stacks up to my birdNET-go install, but it looks promising so far. The fact I was able to get this far is amazing honestly. 

If stranger + AI coding assistances is not what you want helping I totally understand. Happy to help with testing or other things if that is of use. I'm very new to github collaboration, so I'll need to brush up on etiquette if you do think there are ways for me to help, last thing I want is to give you more work because I mess something up.

**Comment by cegan09** [comment]
I haven't gotten to play with it yet, but the CPU based perch V2 was put up finally.

**Comment by cegan09** [comment]
I got it running on a 4gb Pi5, and it seems to work well enough for real time processing as is. I aimed it at a folder of 200 audio clips that are already formatted correctly, and it averages 0.9 clips/s when running in a batch of 2, and 1.13 clips/s in a batch size of 5. I have not tried on a Pi4. I'm going to try to get a fully standalone setup running that intakes two rtsp streams, formats the audio clips, and live evaluates them. I realize that doesn't quite translate to something usable for adding into this project, but it is encouraging that this might run on something like a Pi.

**Reply by tphakala** [reply]
Thanks, nice to hear actual usecase reports. Would you mind to share a quick instructions how you got it running? I haven't had any time to focus on this yet and quick start guide would be much appreciated.

**Reply by cegan09** [reply]
For sure. I need a little time to clean up the larger scripts I'm working on, but the quick points that will prove that you have an environment that works. The notes on the perch model card say you need tensorflow 2.20.0rc0, they don't say minimum, just that it requires that version, I've not yet tested with the official release of 2.20.0 or 2.21.0, just stuck with what they tell you to use. I've done all my testing using tensorflow hub, and had some issues installing both packages without one kicking the version of the other, but running them together with pip install --pre "tensorflow>=2.20.0rc0" tensorflow-hub worked. 

This short script will check that you can load the model, then passes zeros to it just to make sure it runs
```import tensorflow as tf, tensorflow_hub as hub  
url = "https://www.kaggle.com/models/google/bird-vocalization-classifier/frameworks/TensorFlow2/variations/perch_v2_cpu/versions/1"  
m = hub.load(url)  
f = m.signatures.get("serving_default", next(iter(m.signatures.values())))  
inp = list(f.structured_input_signature[1].keys())[0]  
out = f(**{inp: tf.zeros([1,160000], tf.float32)})  
print("OK", {k: v.shape for k,v in out.items()})
```

I'll work on getting the larger scripts up into github today so others can look them over. I have two variants right now, first is just a script you can aim at a folder of sound clips and it will process them, write results to a csv, and report back with some stats on how long it takes to process. The other version is pieces I'm running as services in a larger processing flow. There is one service that accepts an rtsp stream and formats it into the 5second clips that perch wants. Another runs the perch model and batch processes these files as they're coming in and writes the raw results to a local database. And then the messiest one that isn't working quite right is a service that uses the database to comb through and extract the clips with results over some threshold. It also looks for consecutive clips with the same prediction and stitches them together so you don't end up with tons of 5 second clips that likely should be part of the same event. 

I also did some work to add common names to the labels file that perch ships with. By default the perch model just returns an ID number as its prediction, then they give you a labels.csv file with a long list of scientific names that you can match against using the ID as the row number. They provide a second file that gives the ebird 6 digit IDs for any entry that is a bird. I combined all of those and used the inaturalist and ebird published databases to give a single file with scientific and common english names so that I can use that on my end. I'll throw that file up as well, there's probably far cleaner ways to handle that portion than what I did.

**Reply by cegan09** [reply]
I got the first part uploaded. This should have all the required packages you need, the test script to make sure it loads the model correctly, and then a processing script you can aim at a folder of audio clips. I'll keep uploading the other parts as they are ready. 
https://github.com/cegan09/PerchCPU

**Reply by cegan09** [reply]
One last piece of data for now. A friend ran some tests on a pi4 (I'm out of spares right now), and it averages 2.25second/clip at a batch size of 5, which means you can probably run this on a Pi4 if you only have 1 audio stream. More than that might get tight for real time processing. But still encouraging that a pi4 is viable. 

This weekend sometime I should have the rest of my real time processing chain uploaded to that repo.

**Comment by Shellfishgene** [comment]
Note that [here](https://github.com/woheller69/whoBIRD/issues/112#issuecomment-3284349496) Stefan Kahl says the next Birdnet model aims to be better than Perch v2, so depending on how long it will take for that to come out it might not be worth doing the work to implement Perch?

**Reply by cegan09** [reply]
The interesting thing to me with perch is that it does more than birds. I'm mostly using these tools for birds, but it's been really fun having perch catch the coyotes, various frogs, insects, chipmunks, squirrels, etc along side the birds. It's cool to hear that the birdnet model is getting an update though, that should be nice to see.

**Reply by tphakala** [reply]
New BirdNET model has been coming a long time now, I don't believe its existence until it has been released 😄

But BirdNET already supports more species than birds alone, it is trained by various frogs, dogs, coyotes, wolf, etc. I have also been training my own classifier with added species like sheep, red for and moose. https://github.com/tphakala/birdnet-go-classifiers

That being said, I would still like to see Perch model support added, but I haven't done anything about it yet.

**Reply by cegan09** [reply]
I guess I just assumed that with it named BirdNET it was purely bird focused. Sounds like a bad assumption on my part. 
But I largely agree with the idea of supporting multiple models so users can choose the one that works best for them. Perch has been fun to play with, even if I've largely been stumbling in the dark trying to figure it out.

**Comment by uSlackr** [comment]
I think it is bird focused, but as I understand it, part of the issue is that with just a bird model, all sounds are looked at to match a bird.  If other common items are added, you get better bird detection and the added benefit of knowing it was a racoon that just raided your pond 😐

\\Greg
________________________________

**Comment by rambip** [comment]
For my research project, I ended up converting the model to pytorch, is that of any interest ? If so I can share the code and the weights.

**Comment by tariquesani** [comment]
This is exciting news! I don't know how much I will be able to contribute or where, I mostly vibe code with Claude these days, so it might surprise me ;-) I am definitely willing to pitch in
